src/signal.rs

```diff
@@ -474,4 +475,61 @@ mod tests {
         // forever as the loop keeps running. Since we don't join, the thread will become detached
         // as the handle is dropped, and will be killed when the process/main thread exits.
     }
+
+    #[test]
+    fn test_block_unblock_signal() {
+        let signal = SIGRTMIN() + 0;
+
+        // Check if it is blocked.
+        unsafe {
+            let mut sigset: sigset_t = mem::zeroed();
+            pthread_sigmask(SIG_BLOCK, null(), &mut sigset as *mut sigset_t);
+            assert_eq!(sigismember(&sigset, signal), 0);
+        }
+
+        block_signal(signal).unwrap();
+        assert!(get_blocked_signals().unwrap().contains(&(signal)));
+
+        unblock_signal(signal).unwrap();
+        assert!(get_blocked_signals().unwrap().is_empty());
+    }
+
+    #[test]
+    fn test_clear_pending() {
+        unsafe {
+            register_signal_handler(1, SignalHandler::Siginfo(handle_signal), true, SA_SIGINFO)
+                .expect("failed to register vcpu signal handler");
+        }
+
+        // Block the signal, which means it won't be delivered until it is
+        // unblocked. Pending between the time when the signal which is set as blocked
+        // is generated and when is delivered.
+        let killable = thread::spawn(|| {
+            block_signal(SIGRTMIN() + 1).unwrap();
+        });
+
+        // Send a signal to the thread.
+        assert!(killable.kill(1).is_ok());
+
+        // Pending now.
+        unsafe {
+            assert!(!SIGNAL_HANDLER_CALLED[1]);
+        }
+
+        clear_signal(1).unwrap();
```

Comment:
Just wondering, why do we call `block_signal(SIGTRMIN() + 1)`, but later just `clear_signal(1)` (without adding the `SIGTRMIN()`)?

Reply:
Oh, this is a bug of test... And this makes me realize that, it seems hard to really block the signal in sub-thread and kill it... Currently, the block_signal actually didn't work because of not `join`. But once we `join`, the `killable` can not be used any more so can not send the signal... Do you have any suggestion here to really test the clear function with a blocked signal?

Reply:
It seems you pointed me out some problems in the test. `clear_signal` didn't report error when it finds the signal is not pending but returns due to `EAGAIN | EINTR => {}`. 
I did more experiments on the signal handling and updated the test. Per my understanding, `clear_pending()` only removes the pending signal from the list but will not trigger the signal_handler, we could instead test the `sigpending()` finally. Does this make sense?